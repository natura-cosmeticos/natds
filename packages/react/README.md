# Natura Design System - React

React components for web development inside Natura&Co.

[![npm package](https://img.shields.io/npm/v/@naturacosmeticos/natds-react/latest.svg)](https://www.npmjs.com/package/@naturacosmeticos/natds-react)
[![Code coverage](https://img.shields.io/codecov/c/github/natura-cosmeticos/natds-js/main.svg)](https://codecov.io/gh/natura-cosmeticos/natds-js/branch/main)
[![GitHub Issues](https://img.shields.io/github/issues/natura-cosmeticos/natds-js.svg)](https://github.com/natura-cosmeticos/natds-js/issues)
[![License](https://img.shields.io/badge/license-ISC-blue.svg)](https://opensource.org/licenses/ISC)

---

## Getting started

### Requirements

- Please note that `react >= 16.8.4` and `react-dom >= 16.8.4` are peer dependencies.
- We recommend `node >= 8.0.0`
- If your project uses TypeScript, we need `TypeScript >=2.1`;

### Installation

```shell script
$ npm install @naturacosmeticos/natds-web
```

OR

```shell script
$ yarn add @naturacosmeticos/natds-web
```

### Fonts

Load the Roboto font with `400` and `500` font weights:

```html
<link
  href="https://fonts.googleapis.com/css2?family=Roboto+Mono&family=Roboto:wght@400;500&display=swap"
  rel="stylesheet"
/>
```

> This package does not provide Roboto font, only its font family names.

### Icons

Load the font icons from the package `@naturacosmeticos/natds-icons`:

```html
<link
  rel="stylesheet"
  href="https://cdn.jsdelivr.net/npm/@naturacosmeticos/natds-icons@latest/dist/natds-icons.css"
/>
```

or load it from the `node_modules` with a module bundler like `webpack` or `rollup`

`main.js`

```javascript
import React from "react";
import "@naturacosmeticos/natds-icons/natds-icons.css";
```

`webpack.config.js`

```javascript
module.exports = {
  module: {
    loaders: [
      { test: /\.css$/, loader: "style-loader!css-loader" },
      // ...
    ],
  },
};
```

### Usage

Here is a quick example to get you started:

Note that we enclose all your application in a `ThemeProvider` component.

```jsx
import React from "react";
import { Button, ThemeProvider } from "@naturacosmeticos/natds-web";

export const App = () => (
  <ThemeProvider brand="natura" mode="light">
    <Button color="primary" variant="contained">
      Hello World
    </Button>
  </ThemeProvider>
);
```

This package only works fine with the use of `<ThemeProvider />`, that is **essential** for applying styles correctly.

## Additional information

### Bundled in this package

Installing `@naturacosmeticos/natds-react` will also install the following packages:

- `@naturacosmeticos/natds-icons`
- `@naturacosmeticos/natds-styles`

Thus, it is **not necessary to install them separately** to use the Design System.

### Using a CDN

**⚠️ This feature is experimental**.

We provide a Universal Module Definition (UMD) file via `jsdelivr`:

https://cdn.jsdelivr.net/npm/@naturacosmeticos/natds-web

```html
<script
  crossorigin="anonymous"
  src="https://cdn.jsdelivr.net/npm/@naturacosmeticos/natds-web"
></script>
```

For more info and examples, check our [CDN documentation](./docs/using-cdn.md).

---

## Troubleshooting

Having troubles? Please check our [Troubleshooting documentation](../../TROUBLESHOOTING.md).

## Questions

For how-to questions and other issues, please use the [issues section on our GitHub repo](https://github.com/natura-cosmeticos/natds-js/issues).

## Documentation

Check out our [documentation website](https://natds-web.natura.design/).

## Changelog

Recently updated? Please read the [changelog](./CHANGELOG.md).

## License

This project is licensed under the terms of [ISC License](../../LICENSE).
