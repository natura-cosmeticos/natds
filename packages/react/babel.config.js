const path = require('path')

const defaultAlias = {
  '@naturacosmeticos/natds-react': path.resolve(__dirname, './src/components')
}

module.exports = {
  plugins: [
    ['babel-plugin-module-resolver', { alias: defaultAlias }]
  ],
  presets: [
    ['@babel/preset-env', { targets: { node: 'current' } }],
    '@babel/preset-typescript',
    '@babel/preset-react'
  ]
}
