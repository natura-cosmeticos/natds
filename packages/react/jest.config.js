const base = require('../../jest.config.base.js')
const { displayName, name } = require('./package.json')

module.exports = {
  ...base,
  collectCoverageFrom: [
    'src/**/*.{js,jsx,ts,tsx}',
    '!src/**/*.stories.tsx',
    '!src/helpers/**'
  ],
  errorOnDeprecated: true,
  displayName,
  name,
  roots: ['<rootDir>/src'],
  transform: {
    '^.+\\.[tj]sx?$': 'babel-jest',
    '^.+\\.mdx?$': '@storybook/addon-docs/jest-transform-mdx'
  }
}
