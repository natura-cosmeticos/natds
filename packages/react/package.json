{
  "name": "@naturacosmeticos/natds-react",
  "displayName": "NatDS-React",
  "description": "A collection of components from Natura Design System for React",
  "version": "1.8.0-alpha.DSY-1959.4.0",
  "private": false,
  "keywords": [
    "design-system",
    "natura",
    "react",
    "web"
  ],
  "homepage": "https://github.com/natura-cosmeticos/natds-js/tree/master/packages/react/README.md",
  "bugs": {
    "url": "https://github.com/natura-cosmeticos/natds-js/issues"
  },
  "license": "ISC",
  "author": {
    "email": "designsystem@natura.net",
    "name": "Natura&Co.",
    "url": "https://github.com/natura-cosmeticos"
  },
  "files": [
    "README.md",
    "dist/"
  ],
  "jsdelivr": "dist/umd/index.js",
  "source": "src/index.ts",
  "types": "dist/index.d.ts",
  "typings": "dist/index.d.ts",
  "umd:main": "dist/umd/index.js",
  "unpkg": "dist/umd/index.js",
  "main": "dist/index.js",
  "directories": {
    "lib": "dist"
  },
  "repository": {
    "directory": "packages/react",
    "type": "git",
    "url": "git+https://github.com/natura-cosmeticos/natds-js.git"
  },
  "scripts": {
    "build": "yarn clean && tsc -p ./tsconfig.json && yarn rollup:build",
    "build:storybook": "yarn build-storybook -o docs/dist",
    "build:storybook:ci": "yarn build:storybook --quiet",
    "clean": "rm -rf ./dist",
    "copy:assets": "mkdir -p ./dist/assets && cp -r ./src/assets/* ./dist/assets",
    "lint": "eslint --ext .js,.jsx,.ts,.tsx ./src --color",
    "lint:fix": "yarn lint --fix",
    "rollup:build": "rollup -c",
    "storybook": "start-storybook -p 6006",
    "test": "jest --verbose",
    "test:ci": "jest --coverage && codecov -t",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@naturacosmeticos/natds-icons": "1.3.0",
    "@naturacosmeticos/natds-themes": "^0.4.12",
    "react-jss": "10.5.1"
  },
  "devDependencies": {
    "@types/react": "17.0.3",
    "@types/react-dom": "17.0.3",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "react-test-renderer": "17.0.2"
  },
  "peerDependencies": {
    "@types/react": ">= 16.8.6",
    "@types/react-dom": "17.0.3",
    "react": ">= 16.8.4",
    "react-dom": ">= 16.8.4"
  },
  "peerDependenciesMeta": {
    "@types/react": {
      "optional": true
    },
    "@types/react-dom": {
      "optional": true
    }
  }
}
