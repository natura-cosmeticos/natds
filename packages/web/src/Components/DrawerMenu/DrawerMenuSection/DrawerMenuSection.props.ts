import * as React from 'react'
import { IDrawerMenuItemProps } from '../DrawerMenuItem'

export interface IDrawerMenuSectionProps extends IDrawerMenuItemProps {
  list?: [IDrawerMenuSectionProps];
  onToggle?: (event: React.SyntheticEvent, state: boolean, name?: string) => void;
  opened?: boolean;
}
