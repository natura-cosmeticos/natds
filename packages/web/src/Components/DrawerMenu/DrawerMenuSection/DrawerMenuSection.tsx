/* eslint-disable sort-keys */
// eslint-disable-next-line no-use-before-define
import * as React from 'react'
import { createStyles, makeStyles } from '@material-ui/core'
import ListItem from '../../ListItem'
import ListItemIcon from '../../ListItemIcon'
import { Icon } from '../../Icon'
import ListItemText from '../../ListItemText'
import Collapse from '../../Collapse'
import List from '../../List'
import BuildDrawerMenuItems from '../BuildDrawerMenuItems/BuildDrawerMenuItems'
import { MenuItemText } from '../MenuItemText'
import { IDrawerMenuSectionProps } from './DrawerMenuSection.props'
import { IThemeWeb } from '../../../Themes'

export { IDrawerMenuSectionProps } from './DrawerMenuSection.props'

const ICON_ARROW_DOWN = 'outlined-navigation-arrowbottom'
const ICON_ARROW_UP = 'outlined-navigation-arrowtop'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const DrawerMenuSection = React.forwardRef<any, IDrawerMenuSectionProps>((
  props: IDrawerMenuSectionProps, ref
) => {
  const {
    onToggle, icon, name, list, unmountOnExit
  } = props

  const [
    opened, toggleSubmenu
  ] = React.useState(props.opened)

  const useSublistStyles = makeStyles((theme: IThemeWeb) => createStyles({
    padding: {
      borderLeft: `${theme.palette?.action?.hover} 1px solid`,
      margin: `0 0 0 ${theme.sizes?.semiX}px`,
      padding: `0 0 0 ${theme.sizes?.tiny}px`
    }
  }))

  const useListSectionItemStyles = makeStyles((theme: IThemeWeb) => createStyles({
    root: {
      borderRadius: theme.shape?.borderRadius,
      cursor: 'pointer',
      marginBottom: theme.sizes?.micro,
      padding: theme.sizes?.tiny,
      '&:hover': {
        backgroundColor: theme.palette?.action?.hover
      },
      '&$selected': {
        backgroundColor: `${theme.palette?.action?.hover}`,
        '&:hover': {
          backgroundColor: theme.palette?.action?.hover
        }
      }
    },
    selected: {}
  }))

  const useListItemIconStyles = makeStyles((theme: IThemeWeb) => createStyles({
    root: {
      marginRight: theme.sizes?.standard,
      marginTop: 0,
      minWidth: 'auto'
    }
  }))

  const classes = useSublistStyles()
  const classesItem = useListSectionItemStyles()
  const classesIcon = useListItemIconStyles()
  const listIconName = opened ? ICON_ARROW_UP : ICON_ARROW_DOWN

  const handleClick = (event: React.SyntheticEvent) => {
    toggleSubmenu(!opened)

    /**
     * @todo refactor(web): convert expression to if block, add test
     */
    // eslint-disable-next-line no-unused-expressions
    onToggle && onToggle(event, !opened, name)
  }

  return (
    <>
      <ListItem
        onClick={handleClick}
        classes={classesItem}
        selected={opened}
        component="li"
        button
        ref={ref}
      >
        {icon && <ListItemIcon classes={classesIcon}><Icon name={icon} size="tiny" /></ListItemIcon>}
        <ListItemText><MenuItemText {...props} /></ListItemText>
        <Icon name={listIconName} size="tiny" />
      </ListItem>
      <Collapse in={opened} timeout="auto" unmountOnExit={unmountOnExit} component="li">
        {list && (
        <List
          classes={classes}
          dense
        >
          {list.map((drawerMenuSectionProps, index) => (
            <BuildDrawerMenuItems {...drawerMenuSectionProps} key={index} />
          ))}
        </List>
        )}
      </Collapse>
    </>
  )
})

DrawerMenuSection.displayName = 'DrawerMenuSection'

export default DrawerMenuSection
