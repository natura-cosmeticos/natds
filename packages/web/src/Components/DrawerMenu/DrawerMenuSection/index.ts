import DrawerMenuSection from './DrawerMenuSection'

export { IDrawerMenuSectionProps } from './DrawerMenuSection.props'
export { DrawerMenuSection }
export default DrawerMenuSection
