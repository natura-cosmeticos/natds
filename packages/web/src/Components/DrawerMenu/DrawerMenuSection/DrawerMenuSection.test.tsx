import * as React from 'react'
import * as TestRenderer from 'react-test-renderer'
import { DrawerMenuSection } from './DrawerMenuSection'

describe('DrawerMenuSection component', () => {
  it('should match snapshot', () => {
    const testRenderer : TestRenderer.ReactTestRenderer = TestRenderer.create(<DrawerMenuSection />)

    expect(testRenderer).toMatchSnapshot()
  })
})
