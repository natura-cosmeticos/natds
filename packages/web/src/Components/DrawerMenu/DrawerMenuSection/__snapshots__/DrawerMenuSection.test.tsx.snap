// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerMenuSection component should match snapshot 1`] = `
Array [
  <li
    aria-disabled={false}
    className="MuiButtonBase-root MuiListItem-root makeStyles-root-2 MuiListItem-gutters MuiListItem-button"
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="MuiListItemText-root"
    >
      <span
        className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        <span
          className="MuiTypography-root MuiTypography-body2"
        />
      </span>
    </div>
    <i
      aria-hidden={true}
      className="material-icons MuiIcon-root natds-icons natds-icons-outlined-navigation-arrowbottom NatDSIcon-root-6 NatDSIcon-root-7"
    />
  </li>,
  <li
    className="MuiCollapse-root MuiCollapse-hidden"
    style={
      Object {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper"
    >
      <div
        className="MuiCollapse-wrapperInner"
      />
    </div>
  </li>,
]
`;
